### PR TITLE
RWX support and stability improvement

### PR DESCRIPTION
**Pull request by Vicente-Cheng**
I mixed two things because I developed the RWX feature based on the stability improvement.
I apologize for this. I thought the reviewers were the same, so please help to review it!

stability: https://github.com/harvester/harvester-csi-driver/pull/43/commits/f6b6c5a5c34ace0f96b84a8ff92c9b922090af65
RWX: https://github.com/harvester/harvester-csi-driver/pull/43/commits/c28f8af032d6e359af9cef1a33e56bd4419416d2

**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**
- Support LH RWX volume
- Improve the stability of volume operations

**Solution:**
- Support LH RWX volume
- Improve the stability of volume operations

**Related Issue:**
Stability:
- https://github.com/harvester/harvester/issues/5949
- https://github.com/harvester/harvester/issues/5951

RWX:
- https://github.com/harvester/harvester/issues/1992

**Test plan:**

---

### Volume Stability Test Plan
1. Use the following yaml file to deploy 10 volumes on the guest cluster
https://gist.github.com/Vicente-Cheng/8e635a2c0ca649d7256419d61a680b3d
2. Move this workload to another node.

We could increase the volumes to  10, 15, 20, 30, or 40 to ensure stability. 

---

### RWX Support Test Plan
**_Prerequisites_**
- Update the rbac harvester CSI as below
```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
...
  name: harvesterhci.io:csi-driver
...
rules:
...
- apiGroups:
  - harvesterhci.io
  resources:
  - networkfilesystems
  - networkfilesystems/status
  verbs:
  - '*'
- apiGroups:
  - longhorn.io
  resources:
  - volumes
  - volumes/status
  verbs:
  - get
  - list
```

`networkfilesystems` is for the RWX (we also need the new controller (https://github.com/harvester/networkfs-manager)
`volumes` is for the stability checking

- Deploy the `networkfs-manager`
1. Get the chart files on https://github.com/harvester/networkfs-manager/tree/main/deploy/charts
2. Use helm command to install
```
$ helm install harvester-networkfs-manager charts/ -n harvester-system
```

**_Test Steps_**

**_Please complete the `Prerequisites` first_**
1. create 1.4.x harvester cluster
1.1 (Options) create storage network
1.2 (Options) enable RWX volume with storage network on Longhorn (REF: https://longhorn.io/docs/1.7.0/advanced-resources/deploy/storage-network/#setting-storage-network-for-rwx-volumes)
2. create rke2 guest cluster
2.1 (Options) if we enable storage network for RWX volume (on Longhorn), please also add this storage network to the VM for the guest cluster
3. Replace the csi driver image with this PR
4. create longhorn RWX SC on the host cluster.
```
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: longhorn-rwx
provisioner: driver.longhorn.io
allowVolumeExpansion: true
reclaimPolicy: Delete
volumeBindingMode: Immediate
parameters:
  numberOfReplicas: "3"
  staleReplicaTimeout: "2880"
  fromBackup: ""
  fsType: "ext4"
  nfsOptions: "vers=4.2,noresvport,softerr,timeo=600,retrans=5"
```
5. create a new SC associated with the above longhorn-rwx SC as below (on the guest cluster)
```
allowVolumeExpansion: false
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  creationTimestamp: "2024-08-26T06:09:24Z"
  name: longhorn-rwx
  resourceVersion: "98127"
  uid: 697b1e2d-d4fb-4cc1-8ad5-9494c2fd283b
parameters:
  hostStorageClass: longhorn-rwx
provisioner: driver.harvesterhci.io
reclaimPolicy: Delete
volumeBindingMode: Immediate
```
6. create the PVC with this new SC (remember you need to change the access mode to RWX as below)
截圖 2024-08-29 上午10 47 21
7. tested with this PVC should be attached to the multiple pods simultaneously.

### CODE REVIEW

File: pkg/csi/controller_server.go [hunk 1]
```diff
@@ -306,10 +321,27 @@ func (cs *ControllerServer) ControllerUnpublishVolume(_ context.Context, req *cs
 			req.GetVolumeId(), pvc.Status.Phase)
 	}
 
+	volumeHotplugged := false
+	vmi, err := cs.virtClient.VirtualMachineInstance(cs.namespace).Get(context.TODO(), req.GetNodeId(), &metav1.GetOptions{})
```
Comment (bk201) [inline pkg/csi/controller_server.go:356]: can also consider if a VMI is gone

File: pkg/csi/controller_server.go [hunk 2]
```diff
@@ -457,31 +657,111 @@ func getVolumeCapabilityAccessModes(vc []csi.VolumeCapability_AccessMode_Mode) [
...
 			}
-			if predicate(existVol) {
-				return true
+			return true
+		}
+	}
+}
+
+func (cs *ControllerServer) waitForLHVolumeDetachedorAttachedCorrectly(volName, nodeID string) (detached, attached bool) {
+	timeoutTimer := time.NewTimer(timeoutAttachDetach)
```
Comment (bk201) [inline pkg/csi/controller_server.go:690]: nit: might worth checking: https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#PollUntilContextTimeout, while there is nothing wrong implementing the logic by ourself too.

File: pkg/csi/controller_server.go [hunk 3]
```diff
@@ -431,6 +538,99 @@ func (cs *ControllerServer) validateVolumeCapabilities(volumeCaps []*csi.VolumeC
...
+	}
+	pvc.Spec.Resources = corev1.ResourceRequirements{
+		Requests: corev1.ResourceList{
+			corev1.ResourceStorage: *resource.NewQuantity(volSizeBytes, resource.BinarySI),
+		},
+	}
+	return pvc, nil
+}
+
+func (cs *ControllerServer) getVolumeModule(volCaps []*csi.VolumeCapability) corev1.PersistentVolumeMode {
```
Comment (bk201) [inline pkg/csi/controller_server.go:602]: not 100% sure, but feel you might want name it `getVolumeMode`...

File: pkg/csi/controller_server.go [hunk 4]
```diff
@@ -25,27 +30,33 @@ const (
 	timeoutAttachDetach = 120 * time.Second
 	tickAttachDetach    = 2 * time.Second
 	paramHostSC         = "hostStorageClass"
+	LHVolumeNS          = "longhorn-system"
+	HARVNS              = "harvester-system"
```
Comment (bk201) [inline pkg/csi/controller_server.go:34]: Can we use HarvesterNamespace and LonghornNamespace for the constant name for readibility.

File: pkg/csi/controller_server.go [hunk 5]
```diff
@@ -306,10 +349,30 @@ func (cs *ControllerServer) ControllerUnpublishVolume(_ context.Context, req *cs
...
+	}
+	for _, volStatus := range vmi.Status.VolumeStatus {
+		if volStatus.Name == req.VolumeId && volStatus.HotplugVolume != nil {
+			volumeHotplugged = true
+			break
+		}
+	}
+
+	if !volumeHotplugged {
+		logrus.Infof("Volume %s is not attached to node %s. No need RemoVolume operation!", req.GetVolumeId(), req.GetNodeId())
```
Comment (bk201) [inline pkg/csi/controller_server.go:368]:
```suggestion
		logrus.Infof("Volume %s is not attached to node %s. No need RemoveVolume operation!", req.GetVolumeId(), req.GetNodeId())
```

File: pkg/csi/controller_server.go [hunk 6]
```diff
@@ -390,6 +457,46 @@ func (cs *ControllerServer) ControllerExpandVolume(_ context.Context, req *csi.C
 	}, nil
 }
 
+func (cs *ControllerServer) publishRWXVolume(pvc *corev1.PersistentVolumeClaim) (*csi.ControllerPublishVolumeResponse, error) {
+	networkfs, err := cs.harvNetFSClient.HarvesterhciV1beta1().NetworkFilesystems(HARVNS).Get(context.TODO(), pvc.Spec.VolumeName, metav1.GetOptions{})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to delete NetworkFileSystem %s: %v", pvc.Spec.VolumeName, err)
```
Comment (bk201) [inline pkg/csi/controller_server.go:463]:
```suggestion
		return nil, status.Errorf(codes.Internal, "Failed to get NetworkFileSystem %s: %v", pvc.Spec.VolumeName, err)
```

File: pkg/csi/controller_server.go [hunk 7]
```diff
@@ -390,6 +457,46 @@ func (cs *ControllerServer) ControllerExpandVolume(_ context.Context, req *csi.C
...
+	if _, err := cs.harvNetFSClient.HarvesterhciV1beta1().NetworkFilesystems(HARVNS).Update(context.TODO(), networkfs, metav1.UpdateOptions{}); err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to enable NetworkFileSystem %s: %v", pvc.Spec.VolumeName, err)
+	}
+	return &csi.ControllerPublishVolumeResponse{}, nil
+}
+
+func (cs *ControllerServer) unpublishRWXVolume(pvc *corev1.PersistentVolumeClaim) (*csi.ControllerUnpublishVolumeResponse, error) {
+	networkfs, err := cs.harvNetFSClient.HarvesterhciV1beta1().NetworkFilesystems(HARVNS).Get(context.TODO(), pvc.Spec.VolumeName, metav1.GetOptions{})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to delete NetworkFileSystem %s: %v", pvc.Spec.VolumeName, err)
```
Comment (bk201) [inline pkg/csi/controller_server.go:481]:
```suggestion
		return nil, status.Errorf(codes.Internal, "Failed to get NetworkFileSystem %s: %v", pvc.Spec.VolumeName, err)
```

File: pkg/csi/controller_server.go (hunk 2 above)
Comment (bk201) [reply pkg/csi/controller_server.go:690]: How about handling the attach/detach separately (with an argument or different functions) and returning the error from the function?

File: pkg/csi/controller_server.go (hunk 1 above)
Comment (Vicente-Cheng) [reply pkg/csi/controller_server.go:358]: updated!

File: pkg/csi/controller_server.go (hunk 2 above)
Comment (Vicente-Cheng) [reply pkg/csi/controller_server.go]: Thanks for the suggestion, I rewrite this with `PollUnitContextTimeout`.
Please help to check. Thanks!

File: pkg/csi/controller_server.go (hunk 4 above)
Comment (Vicente-Cheng) [reply pkg/csi/controller_server.go:36]: Sure, Update Var to LonghornNS and HarvesterNS

**Review by tserong — COMMENTED** [review]
LGTM, with the disclaimer that I'm not super-familiar with this part of the harvester codebase yet :-)

I wanted to check about the introduction of networkfs-manager. I assume we have this so that in future we can potentially support RWX volumes from non-Longhorn storage backends? My second question about that is: do we also need a PR against harvester to deploy that piece?

I also found https://docs.harvesterhci.io/v1.4/rancher/csi-driver/ which has the following note, which will want updating/removing once this feature goes in:

> Currently, the Harvester CSI driver only supports single-node read-write(RWO) volumes. Please follow the issue #1992 for future multi-node read-only(ROX) and read-write(RWX) support.

**Review by bk201 — APPROVED** [review]
LGTM, thanks!

**Comment by Vicente-Cheng** [comment]
> LGTM, with the disclaimer that I'm not super-familiar with this part of the harvester codebase yet :-)
> 
> I wanted to check about the introduction of networkfs-manager. I assume we have this so that in future we can potentially support RWX volumes from non-Longhorn storage backends? My second question about that is: do we also need a PR against harvester to deploy that piece?

Thanks @tserong,
Yes, the networkfs-manager should be able to handle the external NFS, not only the Longhorn RWX volume.
We will support it in the later version.

You are correct. We need to make some changes on the harvester side, like adding this new component and the new RBAC configuration.
I will file another PR for both things.

> 
> I also found https://docs.harvesterhci.io/v1.4/rancher/csi-driver/ which has the following note, which will want updating/removing once this feature goes in:
> 
> > Currently, the Harvester CSI driver only supports single-node read-write(RWO) volumes. Please follow the issue #1992 for future multi-node read-only(ROX) and read-write(RWX) support.

Indeed, we also need to update the document for detailed examples of users using RWX volume on the downstream clusters.

Thanks!

File: pkg/csi/controller_server.go [hunk 8]
```diff
@@ -457,31 +663,85 @@ func getVolumeCapabilityAccessModes(vc []csi.VolumeCapability_AccessMode_Mode) [
...
+		logrus.Warnf("waitForVolumeSettled: error while waiting for volume %s to be settled. Err: %v", lhVolName, err)
+		return false, err
+	}
+	// check attached correctly or detached
+	if volAttachedCorrectly(volume, nodeID) {
+		return true, nil
+	}
+	if volume.Status.State == lhv1beta2.VolumeStateDetached {
+		return true, nil
+	}
```
Comment (WebberHuang1118) [inline pkg/csi/controller_server.go:678]: The comment is a little confusing, IIUC, `volAttachedCorrectly()` is only for checking volume attachment.

**Review by WebberHuang1118 — APPROVED** [review]
LGTM, thanks for the PR. Will the Harvester chart update with NetworkFS ?

File: pkg/csi/controller_server.go (hunk 8 above)
Comment (Vicente-Cheng) [reply pkg/csi/controller_server.go]: updated, thanks!

**Pull request merged**